lib/selinux/selinux_linux.go

```diff
@@ -0,0 +1,177 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package selinux
+
+import (
+	"bytes"
+	"context"
+	_ "embed"
+	"io/fs"
+	"log/slog"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/gravitational/trace"
+	ocselinux "github.com/opencontainers/selinux/go-selinux"
+
+	"github.com/gravitational/teleport/lib/versioncontrol"
+)
+
+const (
+	selinuxConfig        = "/etc/selinux/config"
+	selinuxTypeTag       = "SELINUXTYPE"
+	moduleName           = "teleport_ssh"
+	domain               = "teleport_ssh_t"
+	permissiveModuleName = "permissive_" + domain
+)
+
+//go:embed teleport_ssh.te
+var module string
+
+//go:embed teleport_ssh.fc.tmpl
+var fileContexts string
+
+// ModuleSource returns the source of the SELinux SSH module.
+func ModuleSource() string {
+	return module
+}
+
+type filePaths struct {
+	InstallDir     string
+	DataDir        string
+	ConfigPath     string
+	UpgradeUnitDir string
+}
+
+// FileContexts returns file contexts for the SELinux SSH module.
+func FileContexts(installDir, dataDir, configPath string) (string, error) {
+	fcTempl, err := template.New("selinux file contexts").Parse(fileContexts)
+	if err != nil {
+		return "", trace.Wrap(err, "failed to parse file contexts template")
+	}
+
+	// Generate a file specifying the locations of important dirs so SELinux
+	// will allow Teleport SSH to be able to access them.
+	var buf bytes.Buffer
+	err = fcTempl.Execute(&buf, filePaths{
+		InstallDir:     installDir,
+		DataDir:        dataDir,
+		ConfigPath:     configPath,
+		UpgradeUnitDir: versioncontrol.UnitConfigDir,
+	})
+	if err != nil {
+		return "", trace.Wrap(err, "failed to expand file contexts template")
+	}
+
+	return buf.String(), nil
+}
+
+// CheckConfiguration returns an error if SELinux is not configured to
+// enforce the SSH service correctly.
+func CheckConfiguration(ensureEnforced bool, logger *slog.Logger) error {
+	if !ocselinux.GetEnabled() {
+		return trace.Errorf("SELinux is disabled or not present")
+	}
+	if ocselinux.EnforceMode() == ocselinux.Disabled {
+		return trace.Errorf("SELinux mode is disabled, SELinux will not enforce or log anything")
+	}
+	if ocselinux.EnforceMode() == ocselinux.Permissive {
+		if ensureEnforced {
+			return trace.Errorf("SELinux mode is permissive, SELinux will not enforce rules only log denials")
+		} else {
+			slog.WarnContext(context.TODO(), "SELinux mode is permissive, SELinux will not enforce rules only log denials")
+		}
+	}
+
+	selinuxType, err := readConfig(selinuxTypeTag)
+	if err != nil {
+		return trace.Wrap(err, "failed to find SELinux type")
+	}
+	selinuxDir := filepath.Join("/var/lib/selinux", selinuxType, "active/modules")
```

Comment:
```suggestion
	selinuxDir := filepath.Join(selinuxRoot, selinuxType, "active", "modules")
```

and add `selinuxRoot` as a const like `selinuxConfig`?